src/proxy/Proxy.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+import { ERC1967Proxy, ERC1967Utils } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import { StorageSlot } from "@openzeppelin/contracts/utils/StorageSlot.sol";
+
+import { Initializable } from "src/attribute/Initializable.sol";
+import { IProxy } from "./interfaces/IProxy.sol";
+import { ProxyAdmin } from "./ProxyAdmin.sol";
+
+/// @title Proxy
+/// @notice A transparent upgradeable proxy with facilities to prevent deployment errors.
+/// @dev The real interface of this proxy is that defined in `IProxy`. This contract does not
+/// inherit from that interface to maintain transparency.  Upgradability is implemented using an
+/// internal dispatch mechanism.  Proxied contracts must implement `Initializable` (not just the
+/// interface) and must return a constant `name` string used for identification.
+/// Implementations must be initialized using calldata during deployment and upgrade.
+contract Proxy is ERC1967Proxy {
```

Comment:
Can we actually just extend the `TransparentUpgradeableProxy` directly?

I think now that our `updateToAndCall` methods has the same signature this should be a really light extension only adding logic to that one method.

I know the note in `TransparentUpgradeableProxy` mentioned not to do this, but given we're only adding extra permissioning onto `updateToAndCall`, this seems like the safer option.

Reply:
As currently implemented, yes, probably.

Reply:
`_dispatchUpgradeToAndCall` is private, not virtual, which means we need to override `_fallback` to call a new _dispatchUpgrade_ method with our logic.

Reply:
This also prevents as from using our own `ProxyAdmin` based on our two-step `Ownable`.  `TransparentUpgradeableProxy` creates it's own admin in its constructor.

Reply:
Also, from a DX perspective, there's value in having our own interfaces enforce the implementation contract subclasses `Initializable`.